### PR TITLE
Popover: refactoring 

### PR DESCRIPTION
**Pull request by retrofox**
In this PT the `<Popover />` component is refactored to trying to fix several bugs that we current have. The main change is that we aren't using `component-tip`.

Issue: #5101, #7003 

Relates bugs: #4588
Remaining issues in this approach:

- [x] setting initial position sometimes fails
- [x] seek and update all Popover instances
- [x] sites-popovers in share - sites are gotten from `lib/sites-list` function. Async issue. autofocus.
- [x] ~~handle multiples popovers~~

image

Let's delegate this task to each integration

Places/pages using `<Popover />` component.

- [x] client/blocks/author-selector/index.jsx
- [x] client/components/chart/bar.jsx
- [x] client/components/info-popover/index.jsx
- [x] client/components/popover/docs/example.jsx
- [x] client/components/popover/menu.jsx
- [x] client/components/plans/premium-popover/index.jsx
- [x] client/components/tooltip/index.jsx
- [x] client/components/tinymce/plugins/contact-form/dialog/field-remove-button.jsx
- [x] client/my-sites/upgrades/cart/popover-cart.jsx
- [x] client/post-editor/edit-post-status/index.jsx
- [x] client/post-editor/editor-ground-control/index.jsx, 
- [x] client/post-editor/editor-ground-control/test/index.jsx
- [x] client/post-editor/editor-visibility/index.jsx
- [x] client/post-editor/test/post-editor.jsx
- [x] client/my-sites/stats/post-trends/day.jsx
- [x] client/post-editor/media-modal/test/index.jsx
- [x] client/components/sites-popover/index.jsx
- [x] client/post-editor/media-modal/gallery-help.jsx

ff - stats - daily

**Test live - Popovers devdocs page**
https://calypso.live/devdocs/design/popovers/?branch=update/popover-refactoring

Test live: https://calypso.live/?branch=update/popover-refactoring

### CODE REVIEW

File: client/blocks/author-selector/index.jsx [hunk 1]
```diff
@@ -82,13 +82,15 @@ SwitcherShell = React.createClass( {
...
+
 				<Popover
+					id='author-selector-popover'
 					isVisible={ this.state.showAuthorMenu }
 					onClose={ this._onClose }
 					position={ this.props.popoverPosition }
 					context={ this.refs && this.refs.authorSelectorChevron }
 					onKeyDown={ this._onKeyDown }
-					className="author-selector__popover popover">
+					className="author-selector__popover popover__container">
```
Comment (mtias) [inline client/blocks/author-selector/index.jsx]: It'd be good if these classes were applied automatically.

File: client/components/chart/bar.jsx [hunk 2]
```diff
@@ -107,11 +107,14 @@ module.exports = React.createClass( {
 		};
 
 		if ( this.props.data.tooltipData && this.props.data.tooltipData.length && ! this.props.isTouch ) {
-			tooltip = <Popover context={ this.refs && this.refs.valueBar }
+			tooltip = <Popover
+							id="popover__chart-bar"
```
Comment (mtias) [inline client/components/chart/bar.jsx]: Why do we need context and id now?

File: client/components/popover/index.jsx [hunk 3]
```diff
@@ -1,160 +1,379 @@
...
+	id: PropTypes.string,
+
+	onClose: PropTypes.func.isRequired,
+	onShow: PropTypes.func,
+};
+
+Popover.defaultProps = {
+	className: 'popover__container',
+	closeOnEsc: true,
+	openDelay: 100,
```
Comment (mtias) [inline client/components/popover/index.jsx]: Why this delay?

File: client/components/popover/style.scss [hunk 4]
```diff
@@ -2,46 +2,38 @@
  * "popover" theme for `component/tip`.
  */
 
-// not sure if we need this
-.tip-hide {
-	opacity: 0;
-}
-
-.popover {
+.popover__container {
```
Comment (mtias) [inline client/components/popover/style.scss]: Why can't this class be just `popover`?

File: client/components/popover/style.scss [hunk 5]
```diff
@@ -85,44 +77,64 @@
...
+	@include popover__arrow( "left" );
+	@include popover__arrow( "right" );
 
-	&.tip-top-left,
-	&.tip-bottom-left,
-	&.tip-top-right,
-	&.tip-bottom-right {
+	&.popover-top-left,
+	&.popover-bottom-left,
+	&.popover-top-right,
```
Comment (mtias) [inline client/components/popover/style.scss]: All these classes should be changed to `is-top-left` etc

File: client/components/sites-popover/index.jsx [hunk 6]
```diff
@@ -21,6 +21,7 @@ module.exports = React.createClass( {
 		visible: React.PropTypes.bool,
 		onClose: React.PropTypes.func,
 		position: React.PropTypes.string,
+		id: React.PropTypes.string,
```
Comment (mtias) [inline client/components/sites-popover/index.jsx]: Again, I think we should infer this id or get rid of context.

**Comment by mtias** [comment]
cc @ockham who I think looked at a react-popover component a while ago.

**Comment by retrofox** [comment]
- [x] client/blocks/author-selector/index.jsx

image

**Comment by retrofox** [comment]
- [x] client/components/chart/bar.jsx

image

https://cloudup.com/iOdDiwX11cj

File: client/components/popover/index.jsx (hunk 3 above)
Comment (retrofox) [inline client/components/popover/index.jsx]: Sorry by the delay :-p

This parameter, renamed to `showDelay`, add debounce when the Popover is opened. It optimizes some edge cases when we have a lot of targets all together.

Although it's outside my area I think also it improves the UX.
https://github.com/Automattic/wp-calypso/pull/7217#issuecomment-237300078

Anyway this parameter is optional and the default options isn't add the delay.

File: client/components/chart/bar.jsx (hunk 2 above)
Comment (retrofox) [inline client/components/chart/bar.jsx]: This `id` is optional not needed. For now I'm using this `id` JUST for debugging.

image

image

it's a mess but for now we don't have an isolation environment to check.

Also it could be useful to try to avoid creating multiple popovers when you could use only one for a bunch of targets but I haven't done anything to trying to tackle this scenario.

File: client/components/popover/style.scss (hunk 4 above)
Comment (retrofox) [inline client/components/popover/style.scss]: I'm getting the eslint-wanring `className should follow CSS name ..._prefix`. I guess it's because `<Popover />` is child of `<RootChild />` component.

image

We can easily avoid this warning using `classNames()` function. (In fact it's already used in this PR)

**Comment by retrofox** [comment]
- [x] client/components/info-popover/index.jsx

image

**Comment by retrofox** [comment]
premium-popover

**before (right now in prod)**
image

**Comment by retrofox** [comment]
**`<Tooltip />`**

- [x] client/components/tooltip/index.jsx
- [x] added example to `devdocs` page -> http://calypso.localhost:3000/devdocs/design/tooltip

image

**Comment by retrofox** [comment]
- [x] client/components/tinymce/plugins/contact-form/dialog/field-remove-button.jsx

image

**Comment by retrofox** [comment]
- [x] client/my-sites/upgrades/cart/popover-cart.jsx

image

**Comment by retrofox** [comment]
- [ ] client/post-editor/editor-ground-control/index.jsx
- [ ] client/post-editor/editor-ground-control/test/index.jsx
- [ ] client/post-editor/editor-visibility/index.jsx
- [ ] client/post-editor/media-modal/gallery-help.jsx
- [ ] client/post-editor/media-modal/test/index.jsx
- [ ] client/post-editor/test/post-editor.jsx

image

image

image

**Comment by retrofox** [comment]
- [x] client/post-editor/media-modal/gallery-help.jsx

image

**Comment by retrofox** [comment]
cc @eliorivero

**Comment by retrofox** [comment]
cc @coderkevin

**Comment by gwwar** [comment]
@retrofox this changeset is quite large, so you may need to ping folks per section to verify that no behavior has changed.

**Comment by retrofox** [comment]
> @retrofox this changeset is quite large, so you may need to ping folks per section to verify that no behavior has changed.

I'm afraid you're right. Let me make a simple search an create an assignation list.

**Comment by retrofox** [comment]
Hi developers. I'm ping you in order to get your support with this review. The idea is get your feedback of after this refactoring. To build this list I did a `git blabe` search for each file and in some cases the searching hasn't been perfect. Feel free either to take a look and check your checkbox or auto-remove your name form the list.

### Assignment list

- [ ] client/blocks/author-selector/index.jsx (_volunteer_)
- [ ] client/blocks/author-selector/style.scss (_volunteer_)
- [ ] client/components/chart/bar.jsx (_volunteer_)
- [ ] client/components/chart/style.scss (_volunteer_)
- [ ] client/components/info-popover/docs/example.jsx (_volunteer_)
- [ ] client/components/info-popover/index.jsx (_volunteer_)
- [ ] client/components/info-popover/style.scss (_volunteer_)
- [ ] client/components/plans/premium-popover/index.jsx (@umurkontaci)
- [ ] client/components/plans/premium-popover/style.scss (@umurkontaci)
- [ ] client/components/popover/docs/example.jsx (_volunteer_)
- [ ] client/components/popover/index.jsx (_volunteer_)
- [ ] client/components/popover/style.scss (_volunteer_)
- [ ] client/components/popover/util.js (_volunteer_)
- [ ] client/components/post-schedule/style.scss (_me_)
- [ ] client/components/sites-popover/index.jsx (@blowery)
- [ ] client/components/sites-popover/style.scss (@blowery)
- [ ] client/components/tinymce/plugins/contact-form/dialog/field-remove-button.jsx (@rodrigoi)
- [ ] client/components/tooltip/docs/example.jsx (_volunteer_)
- [ ] client/components/tooltip/index.jsx (_volunteer_)
- [ ] client/components/tooltip/style.scss (_volunteer_)
- [ ] client/devdocs/design/index.jsx (_me_)
- [ ] client/devdocs/style.scss (_me_)
- [ ] client/layout/masterbar/publish.jsx (@folletto)
- [ ] client/my-sites/upgrades/cart/popover-cart.jsx (@ockham)
- [ ] client/post-editor/editor-ground-control/index.jsx (@coreh)
- [ ] client/post-editor/editor-visibility/index.jsx (@jeremeylduvall)
- [ ] client/post-editor/editor-visibility/style.scss (@jeremeylduvall)
- [ ] client/post-editor/media-modal/gallery-help.jsx (@aduth)
- [ ] client/post-editor/media-modal/style.scss (@aduth)

image

**Comment by folletto** [comment]
> Feel free either to take a look and check your checkbox or auto-remove your name form the list.

Can you outline _exactly_ what has to be done to check the item from the list?

**Comment by retrofox** [comment]
yes, sure and sorry. yes, sure. You should only check correct functionality of Popover trying to generate some edge case. That's all. Thanks @folletto.

**Comment by rodrigoi** [comment]
I can confirm that the _Remove Field_ button on the Contact Form Dialog works as expected.

However, I do get some warnings then compiling the CSS files:

```
WARNING: No layer found for `.popover` of `[.popover, .input-chrono__container .gridicons-calendar]` in $z-layers map. Property omitted.
Backtrace:
	assets/stylesheets/shared/functions/_z-index.scss:217, in function `map-deep-get`
	assets/stylesheets/shared/functions/_z-index.scss:226, in function `z-index`
	client/components/input-chrono/style.scss:7

WARNING: No layer found for `.popover` of `[.popover, input.input-chrono]` in $z-layers map. Property omitted.
Backtrace:
	assets/stylesheets/shared/functions/_z-index.scss:217, in function `map-deep-get`
	assets/stylesheets/shared/functions/_z-index.scss:226, in function `z-index`
	client/components/input-chrono/style.scss:29
```

**Comment by folletto** [comment]
> /client/layout/masterbar/publish.jsx

This uses `SitesPopover` and it's working correctly in Chrome, Safari, Firefox:

screen shot 2016-08-11 at 20 33 01

**Comment by retrofox** [comment]
I'll check your box for you @folletto :-D

**Comment by retrofox** [comment]
thank you very much @rodrigoi. Fixed in 1ef835f

**Comment by retrofox** [comment]
- [ ] It'd be good if the popover classes were applied automatically.
- [ ] removed chart/tooltip component and use <Tooltip /> component.

**Comment by retrofox** [comment]
@rralian I've done the changes that we talked last week. In short set `popover` css class for the component and add this one automatically so in this way parent components don't need add it manually.

When I did this I had to refactor a Tooltip component defined into components/chart/ component. it wasn't the same component into components/tooltip. The good news is We've removed duplicated code since we had two Tooltips components doind the same. The bad news is this PR continues growing up.

I'd like merge this one and then we can create new issues if it's needed.

To merge this PR we will have to check all places where the Popover/Tooltip component is used.

File: client/post-editor/media-modal/gallery-help.jsx [hunk 7]
```diff
@@ -80,7 +80,7 @@ const EditorMediaModalGalleryHelp =  React.createClass( {
 				context={ renderContext }
 				position="bottom"
 				isVisible={ ! isMobile() }
-				className="popover is-dialog-visible">
+				className="popover popover__gallery-help is-dialog-visible">
```
Comment (rralian) [inline client/post-editor/media-modal/gallery-help.jsx]: I notice you're adding a new class here that doesn't seem related to the other changes in this commit. Just checking, is this something we need?

**Comment by gwwar** [comment]
@retrofox there's a missed reference to 'components/chart/tooltip' in https://github.com/Automattic/wp-calypso/blob/update/popover-refactoring/client/my-sites/stats/post-trends/day.jsx#L13

**Comment by rralian** [comment]
> @retrofox there's a missed reference to 'components/chart/tooltip' in

@gwwar the latest commit fixes it.

**Comment by gwwar** [comment]
When I'm scheduling a post, the popover gets into a bad state the second time it's opened:
screen shot 2016-08-15 at 3 55 56 pm

1. Navigate to http://calypso.localhost:3000/post
2. Select a site
3. Write something
4. Click on "Set Date and Time" (The calendar next to publish )
5. Calendar pops up, you can interact with it as normal
6. Dismiss it by clicking else where
7.  Click on "Set Date and Time"
8. Any clicks on the calendar dismiss the popup

**Comment by gwwar** [comment]
The following tip points at the wrong item:
screen shot 2016-08-15 at 3 50 36 pm

1. Navigate to http://calypso.localhost:3000/post
2. Select a site
3. Hover over these icons in the top right (should work properly)
4. Resize window
5. Hover over the sticky icon. It points at the Visibility Icon

**Comment by gwwar** [comment]
The cart is visually incorrect, and can't be interacted with the second time it's opened
screen shot 2016-08-15 at 4 05 35 pm

1. Navigate to /plans
2. Click on Upgrade (for one of the plans)
3. Navigate back to /plans
4. Click on the cart icon
5. Close the cart
6. Click on the cart icon again
7. Any clicks on the cart dismiss the popup

**Comment by retrofox** [comment]
@gwwar thanks for the review.

> When I'm scheduling a post, the popover gets into a bad state the second time it's opened:

> The cart is visually incorrect, and can't be interacted with the second time it's opened

I put a race-condition that produced this bug and it was present for all `<Popover />` instances that we were creating.

First (yesterday) the clickout-side event was bound/unbound watching the visibility. Try to track this state became a little bit confused and tough.

Then (last night) the event was bound/unbound in the respective show/hide methods of the component. It was ok but I had to hack the call adding a timeout to try to ensure that the dom element was right there.

And it fired the first step of the solution. And a lesson that I had to learn: "A real life isn't virtual. A real DOM either".
To ensure that the real DOM is right there we have to use a callback function defined in the ref property of the component:

https://facebook.github.io/react/docs/more-about-refs.html#the-ref-callback-attribute

```es6
this.setDomBahevior( el ) {
  if ( el ) {
    this.bindClickoutSide( el );
  else {
    this.unbindClickoutSide();
  }
}

render() {
  return(
    <div
      ref="setDOMBehavior"
     // ...
    >
      { this.props.children }
    </div>
  );
}
```

And it was all at least for this issue ... for now :-o

File: client/post-editor/editor-action-bar/style.scss [hunk 8]
```diff
@@ -39,3 +39,7 @@
 		color: darken( $gray, 10% );
 	}
 }
+
+.editor-action-bar__view-post-tooltip .popover__inner {
+	white-space: nowrap;
+}
```
Comment (gwwar) [inline client/post-editor/editor-action-bar/style.scss]: needs a newline

File: client/post-editor/editor-delete-post/style.scss [hunk 9]
```diff
@@ -8,3 +8,7 @@
 		animation: loading-dot-pulse 0.8s ease-in-out infinite;
 	}
 }
+
+.editor-delete-post__tooltip .popover__inner {
+	white-space: nowrap;
+}
```
Comment (gwwar) [inline client/post-editor/editor-delete-post/style.scss]: needs a newline

File: client/post-editor/editor-sticky/style.scss [hunk 10]
```diff
@@ -8,3 +8,7 @@
 	display: block;
 	font-size: 11px;
 }
+
+.editor-sticky__tooltip .popover__inner {
+	white-space: nowrap;
+}
```
Comment (gwwar) [inline client/post-editor/editor-sticky/style.scss]: needs a newline

File: client/devdocs/style.scss [hunk 11]
```diff
@@ -211,3 +211,48 @@
...
+
+	&:nth-child(4n + 2) {
+		background-color: #ffffe0;
+
+		&:hover {
+			background-color: $gray;
+			color: $white;
+		}
+	}
+}
```
Comment (gwwar) [inline client/devdocs/style.scss]: end of line

**Comment by gwwar** [comment]
This one isn't a blocker, but for tall bars, we should change the tooltip direction
screen shot 2016-08-16 at 10 18 21 am

**Comment by gwwar** [comment]
Alright, looking good @retrofox! I think most of the main paths have been tested. Let's get this in and clean up any missed regressions in follow up PRs.

**Comment by retrofox** [comment]
Thanks for your awesome review.

**Comment by blowery** [comment]
addressed one issue stemming from this change in https://github.com/Automattic/wp-calypso/pull/7883

**Comment by ebinnion** [comment]
It looks like this PR might have introduced a regression on the edit user page `/people/edit/$username/$site` route.

When clicking select a user, the `AuthorSelector` popover should show. But, in current master it does not show.

screen shot 2016-12-08 at 12 03 01 am

**Comment by retrofox** [comment]
thanks @ebinnion for this report. Please let me know if I can help you. Please ping me in your PR.
